visualization/app/codeCharta/ui/searchPanelModeSelector/searchPanelModeSelector.po.ts

```diff
@@ -4,8 +4,8 @@ import { delay } from "../../../puppeteer.helper"
 export class SearchPanelModeSelectorPageObject {
 	constructor(private page: Page) {}
 
-	public async openTreeView() {
-		await this.page.click("search-panel-mode-selector-component #tree-view")
+	public async toggleTreeView() {
+		await this.page.click("#tree-view")
 		await delay(500)
```

Comment:
Suggestion: it might be useful to return the state (if the tree view is now enabled or disabled) as boolean or string. That way it's possible to check that in the test.

Reply:
Instead of returning a boolean here, I added a new function `isTreeViewOpen`.
It was not tested before if `toggleTreeView` actually works (I just saw it does when running the tests headful), so I added a test for this function now.

Reply:
Thanks, that's definitely an improvement. The reason why I would still implement it with the return value though is that it's not guaranteed that the test conditions could in theory differ in a way that the new `toggleTreeView` would work as expected but that it works differently under other circumstances.

A reason for the current implementation might be that it's not that obvious how to track what page clicked how many times. That is possible by using a `WeakMap`. The key is the page and the value is the number of clicks. That way it's possible to know if it should be open or closed. And it would of course also be possible to wait for the addition or removal of the `current` class from `#tree-view` to remove the delay.

That's just a suggestion though and I am fine with it is as well.

Reply:
I think think that checking if the tree-view is open or not by simply checking if the classNames include `current` is way more convinient than keeping track of the current clicks in a map.

```
public async isTreeViewOpen() {
	const classNames = await this.page.$eval("#tree-view", el => el.className)

	return classNames.includes("current")
}
```

This has been implemented already.

I'll get rid of the delays in another PR. The next PR will introduce jest-puppeteer and speed things up with a different setup and teardown. After that I'll replace all delays with `waitForSelector()` calls.